solidity/contracts/ERC20.sol

```diff
@@ -0,0 +1,312 @@
+	* @dev See {IERC20-approve}.
+	*
+	* NOTE: If `value` is the maximum `uint256`, the allowance is not updated on
+	* `transferFrom`. This is semantically equivalent to an infinite approval.
+	*
+	* Requirements:
+	*
+	* - `spender` cannot be the zero address.
+	*/
+	function approve(address spender, uint256 value) public virtual returns (bool) {
+		address owner = _msgSender();
+		_approve(owner, spender, value);
+		return true;
+	}
+
+	/**
+	* @dev See {IERC20-transferFrom}.
+	*
+	* Skips emitting an {Approval} event indicating an allowance update. This is not
+	* required by the ERC. See {xref-ERC20-_approve-address-address-uint256-bool-}[_approve].
+	*
+	* NOTE: Does not update the allowance if the current allowance
+	* is the maximum `uint256`.
+	*
+	* Requirements:
+	*
+	* - `from` and `to` cannot be the zero address.
+	* - `from` must have a balance of at least `value`.
+	* - the caller must have allowance for ``from``'s tokens of at least
+	* `value`.
+	*/
+	function transferFrom(address from, address to, uint256 value) public virtual returns (bool) {
+		address spender = _msgSender();
+		_spendAllowance(from, spender, value);
+		_transfer(from, to, value);
+		return true;
+	}
+
+	/**
+	* @dev Moves a `value` amount of tokens from `from` to `to`.
+	*
+	* This internal function is equivalent to {transfer}, and can be used to
+	* e.g. implement automatic token fees, slashing mechanisms, etc.
+	*
+	* Emits a {Transfer} event.
+	*
+	* NOTE: This function is not virtual, {_update} should be overridden instead.
+	*/
+	function _transfer(address from, address to, uint256 value) internal {
+		if (from == address(0)) {
+			revert ERC20InvalidSender(address(0));
+		}
+		if (to == address(0)) {
+			revert ERC20InvalidReceiver(address(0));
+		}
+		_update(from, to, value);
+	}
+
+	/**
+	* @dev Transfers a `value` amount of tokens from `from` to `to`, or alternatively mints (or burns) if `from`
+	* (or `to`) is the zero address. All customizations to transfers, mints, and burns should be done by overriding
+	* this function.
+	*
+	* Emits a {Transfer} event.
+	*/
+	function _update(address from, address to, uint256 value) internal virtual {
+		if (from == address(0)) {
+			// Overflow check required: The rest of the code assumes that totalSupply never overflows
+			_totalSupply += value;
+		} else {
+			uint256 fromBalance = _balances[from];
+			if (fromBalance < value) {
+				revert ERC20InsufficientBalance(from, fromBalance, value);
+			}
+			unchecked {
+				// Overflow not possible: value <= fromBalance <= totalSupply.
+				_balances[from] = fromBalance - value;
+			}
+		}
+
+		if (to == address(0)) {
+			unchecked {
+				// Overflow not possible: value <= totalSupply or value <= fromBalance <= totalSupply.
+				_totalSupply -= value;
+			}
+		} else {
+			unchecked {
+				// Overflow not possible: balance + value is at most totalSupply, which we know fits into a uint256.
+				_balances[to] += value;
+			}
+		}
+
+		emit Transfer(from, to, value);
```

Comment:
*grumble* Have I lost this battle, has the entire ecosystem settled on mints just being transfer from 0 and burns being transfer to 0, rather than having a proper `mint` and `burn` event like a reasonable species?